protocol/messenger_messages_order_controller_test.go

```diff
@@ -0,0 +1,126 @@
+package protocol
+
+import (
+	"bytes"
+	"sort"
+	"sync"
+
+	"github.com/status-im/status-go/eth-node/types"
+	"github.com/status-im/status-go/protocol/transport"
+)
+
+type messagesOrderType int
+
+const (
+	messagesOrderRandom messagesOrderType = iota
+	messagesOrderInPostOrder
+	messagesOrderReversed
+)
+
+type MessagesOrderController struct {
+	order               messagesOrderType
+	messagesInPostOrder [][]byte
+	mutex               sync.RWMutex
+	quit                chan struct{}
+	quitOnce            sync.Once
+}
+
+func NewMessagesOrderController(order messagesOrderType) *MessagesOrderController {
+	return &MessagesOrderController{
+		order:               order,
+		messagesInPostOrder: [][]byte{},
+		mutex:               sync.RWMutex{},
+		quit:                make(chan struct{}),
+	}
+}
+
+func (m *MessagesOrderController) Start(c chan *PostMessageSubscription) {
+	go func() {
+		for {
+			select {
+			case sub, more := <-c:
+				if !more {
+					return
+				}
+				m.mutex.Lock()
+				m.messagesInPostOrder = append(m.messagesInPostOrder, sub.id)
+				m.mutex.Unlock()
+
+			case <-m.quit:
+				return
+			}
+		}
+	}()
+}
+
+func (m *MessagesOrderController) Stop() {
+	m.quitOnce.Do(func() {
+		close(m.quit)
+	})
+}
+
+func (m *MessagesOrderController) newMessagesIterator(chatWithMessages map[transport.Filter][]*types.Message) MessagesIterator {
+	switch m.order {
+	case messagesOrderInPostOrder, messagesOrderReversed:
+		return &messagesIterator{chatWithMessages: m.sort(chatWithMessages, m.order)}
+	}
+
+	return NewDefaultMessagesIterator(chatWithMessages)
+}
+
+func indexOf(slice [][]byte, value []byte) int {
+	for i, v := range slice {
+		if bytes.Equal(v, value) {
+			return i
+		}
+	}
+	return -1 // Not found
+}
+
+func (m *MessagesOrderController) sort(chatWithMessages map[transport.Filter][]*types.Message, order messagesOrderType) []*chatWithMessage {
+	allMessages := make([]*chatWithMessage, 0)
+	for chat, messages := range chatWithMessages {
+		for _, message := range messages {
+			allMessages = append(allMessages, &chatWithMessage{chat: chat, message: message})
+		}
+	}
+
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	sort.SliceStable(allMessages, func(i, j int) bool {
+		indexI := indexOf(m.messagesInPostOrder, allMessages[i].message.Hash)
+		indexJ := indexOf(m.messagesInPostOrder, allMessages[j].message.Hash)
```

Comment:
Very minor performance things, but anyway.
This will lead to call `indexOf` multiple times for the same index, right? Maybe we could first calculate indexes for all hashes and then `sort.SliceStable`?

Reply:
not a big deal, the number of message is usually very small in tests anyway.

Reply:
Addressed: https://github.com/status-im/status-go/compare/55f13250ea817a2ffe37889a38a5be36b2cadc3d..82ae52b73d0646bd7a74b127d86c967fc8998e27